tests/unit/dataactvalidator/test_d43_detached_award_financial_assistance_2.py

```diff
@@ -0,0 +1,97 @@
+from dataactcore.models.domainModels import Zips
+
+_FILE = 'd43_detached_award_financial_assistance_2'
+
+
+def test_column_headers(database):
+    expected_subset = {"row_number", "place_of_performance_zip4a", "place_of_performance_congr",
+                       "place_of_performance_code"}
+    actual = set(query_columns(_FILE, database))
+    print(actual)
+    assert expected_subset == actual
+
+
+def test_success(database):
+    """ When provided, PrimaryPlaceofPerformanceZIP+4, placeOfPerformanceCongressCode must be in the state specified
+        by PrimaryPlaceOfPerformanceCode. Ignore cases where zip4 isn't provided """
+
+    zips1 = Zips(zip5="12345", zip_last4="6789", state_abbreviation="NY", congressional_district_no="04")
+    zips2 = Zips(zip5="12345", zip_last4="6780", state_abbreviation="NY", congressional_district_no="03")
+    zips3 = Zips(zip5="12345", zip_last4="6781", state_abbreviation="NY", congressional_district_no="02")
+    zips4 = Zips(zip5="12345", zip_last4="6782", state_abbreviation="NY", congressional_district_no="01")
+    # ignored because no zip4
+    det_award_1 = DetachedAwardFinancialAssistanceFactory(place_of_performance_zip4a="",
+                                                          place_of_performance_congr="01",
+                                                          place_of_performance_code="NY*****")
+
+    det_award_2 = DetachedAwardFinancialAssistanceFactory(place_of_performance_zip4a=None,
+                                                          place_of_performance_congr="03",
+                                                          place_of_performance_code="NY*****")
+    # valid 5 digit zip
+    det_award_3 = DetachedAwardFinancialAssistanceFactory(place_of_performance_zip4a="12345",
+                                                          place_of_performance_congr="04",
+                                                          place_of_performance_code="NY*****")
+
+    det_award_4 = DetachedAwardFinancialAssistanceFactory(place_of_performance_zip4a="12345",
+                                                          place_of_performance_congr="01",
+                                                          place_of_performance_code="NY*****")
+    # valid 9 digit zip
+    det_award_5 = DetachedAwardFinancialAssistanceFactory(place_of_performance_zip4a="123456789",
+                                                          place_of_performance_congr="04",
+                                                          place_of_performance_code="NY*****")
+
+    det_award_6 = DetachedAwardFinancialAssistanceFactory(place_of_performance_zip4a="123456780",
+                                                          place_of_performance_congr="03",
+                                                          place_of_performance_code="NY*****")
+
+    det_award_7 = DetachedAwardFinancialAssistanceFactory(place_of_performance_zip4a="12345-6781",
```

Comment:
You don't have to check for all types of zips, that's not what this rule is testing at all. What we need to look at is different district numbers for different states (upper and lowercase) and make sure it's validating all the cases. Same comment for failures